pkg/admission/config.go

```diff
@@ -0,0 +1,117 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package admission
+
+import (
+	"fmt"
+	"io/ioutil"
+
+	"github.com/ghodss/yaml"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/apis/componentconfig"
+	componentconfigv1alpha1 "k8s.io/kubernetes/pkg/apis/componentconfig/v1alpha1"
+	"k8s.io/kubernetes/pkg/runtime"
+)
+
+// ReadAdmissionConfiguration reads the admission configuration at the specified path.
+// It returns the loaded admission configuration if the input file aligns with the required syntax.
+// If it does not align with the provided syntax, it returns a default configuration.
+// It does this to preserve backward compatibility when admission control files were opaque.
+// It returns an error if the file did not exist.
+func ReadAdmissionConfiguration(configFilePath string) (*componentconfig.AdmissionConfiguration, error) {
+	externalConfig := &componentconfigv1alpha1.AdmissionConfiguration{}
+	internalConfig := &componentconfig.AdmissionConfiguration{}
+
+	// a file was provided, so we just read it.
+	if configFilePath != "" {
+		data, err := ioutil.ReadFile(configFilePath)
+		if err != nil {
+			return internalConfig, fmt.Errorf("unable to read admission control configuration from %q [%v]", configFilePath, err)
+		}
+		// if we are unable to decode, we assume this is legacy input that is unable to do multiple opaque configurations.
+		if err := runtime.DecodeInto(api.Codecs.UniversalDecoder(), data, externalConfig); err != nil {
+			return internalConfig, nil
```

Comment:
Not seeing how internalConfig is populated in this case.

It might make more sense to just try a normal Decode first, which would automatically apply defaults and conversions:
```
decoder := api.Codecs.UniversalDecoder()
decodedObj, err := runtime.Decode(decoder, data)
if err == nil {
	decodedConfig, ok := decodedObj.(*componentconfig.AdmissionConfiguration)
	if !ok {
		return nil, fmt.Errorf("unexpected type: %T", decodedObj)
	}
	return decodedConfig, nil
}

```

or return an error if the decode error wasn't related to a missing type:
```
if !(runtime.IsMissingVersion(err) || runtime.IsMissingKind(err) || runtime.IsNotRegisteredError(err)) {
	return nil, err
}
```

finally, special case the conversion of the old unversioned format:
* call DecodeInto into an object matching the old admission file schema (or just json decode into map[string]interface{})
* transform the resulting object into a v1alpha1.AdmissionConfiguration object
* apply defaults
* convert to internal.AdmissionConfiguration